File: rust-sdk/tests/sync_test.rs
//! sync（prod→stg フル複製）のテスト（設計 §4.2・TASK-50 C4）。
//!
//! `KijukuDB::replicate_db` が prod(RO) → stg(RW) の完全な複製を行うこと、
//! 既存 stg の上書きが冪等であること、src==dst を誤設定として弾くことを検証する。

use kijuku_db::{KijukuDB, MediaInput, MediaType};
use tempfile::TempDir;

/// 末尾に2件のメディアを持つ prod DB を作成し、パスを返す。
fn setup_prod(dir: &TempDir) -> (std::path::PathBuf, std::path::PathBuf) {
    let prod = dir.path().join("kijuku.db");
    let stg = dir.path().join("kijuku.stg.db");
    let prod_db = KijukuDB::open(&prod).expect("open prod");
    prod_db.migrate().expect("migrate prod");
    prod_db
        .create_media(&MediaInput {
            title: "prod メディア1".to_string(),
            media_type: MediaType::Comic,
            ..Default::default()
        })
        .expect("create media 1");
    prod_db
        .create_media(&MediaInput {
            title: "prod メディア2".to_string(),
            media_type: MediaType::Comic,
            ..Default::default()
        })
        .expect("create media 2");
    (prod, stg)
}

/// prod(RO)→stg(RW) のフル複製がデータ・スキーマ・バージョンを正しくコピーする（設計 §4.2）。
#[test]
fn test_replicate_db_copies_data_and_schema() {
    let dir = TempDir::new().unwrap();
    let (prod, stg) = setup_prod(&dir);

    let (prod_version, prod_tables) = {
        let prod_db = KijukuDB::open(&prod).unwrap();
        let v = prod_db.get_schema_version().unwrap();
        let t = prod_db.get_tables().unwrap();
        (v, t)
    };
    assert!(!stg.exists(), "sync 前は stg が存在しない");

    // sync（prod→stg）
    KijukuDB::replicate_db(&prod, &stg).unwrap();

    // stg を開いて prod と一致することを検証
    let stg_db = KijukuDB::open(&stg).unwrap();
    assert_eq!(stg_db.get_schema_version().unwrap(), prod_version);
    assert_eq!(stg_db.get_tables().unwrap(), prod_tables);
    assert_eq!(stg_db.get_media(1).unwrap().title, "prod メディア1");
    assert_eq!(stg_db.get_media(2).unwrap().title, "prod メディア2");
}

/// 既存 stg（stg 側の追加分を含む）がある状態での再 sync が、prod で完全に上書きされる（設計 §4.2）。
#[test]
fn test_replicate_db_overwrites_existing_stg() {
    let dir = TempDir::new().unwrap();
    let (prod, stg) = setup_prod(&dir);

    // 1回目の sync
    KijukuDB::replicate_db(&prod, &stg).unwrap();
    assert!(stg.exists());

    // stg 側で独自にレコードを追加（LLM 編集のシミュレート）
    {
        let stg_db = KijukuDB::open(&stg).unwrap();
        stg_db
            .create_media(&MediaInput {
                title: "stg 側の追加分（再 sync で消えるべき）".to_string(),
                media_type: MediaType::Comic,
                ..Default::default()
            })
            .expect("stg append");
    }

    // 再 sync（既存 stg + WAL/SHM 副産物が残り得る状態からの上書き）
    KijukuDB::replicate_db(&prod, &stg).unwrap();

    // 再 sync 後: stg は prod と完全一致（stg 固有の追加分は削除され、prod データのみ残る）
    let stg_db = KijukuDB::open(&stg).unwrap();
    assert_eq!(stg_db.get_media(1).unwrap().title, "prod メディア1");
    assert_eq!(stg_db.get_media(2).unwrap().title, "prod メディア2");
    assert!(
        stg_db.get_media(3).is_none(),
        "再 sync で stg 固有の追加分は prod で上書きされて削除される"
    );
}

/// src == dst は誤設定としてエラー（自己コピー防止・設計 §4.2）。
#[test]
fn test_replicate_db_rejects_same_path() {
    let dir = TempDir::new().unwrap();
    let same = dir.path().join("kijuku.db");
    let result = KijukuDB::replicate_db(&same, &same);
    assert!(result.is_err(), "src == dst はエラー");
    let err = result.unwrap_err().to_string();
    assert!(
        err.contains("same path"),
        "expected same-path error, got: {}",
        err
    );
}

/// discard（stg 破棄・再 sync・設計 §4.6・TASK-45）: gate 不合格等で stg を捨てて prod から
/// 再構築する経路。処理は sync と同一（`replicate_db(prod, stg)`）のため、ここでは discard の
/// 意味論（stg の編集破棄 + prod 無傷）を検証する。CLI/remote の `discard` operation も同一処理。
#[test]
fn test_discard_restores_stg_from_prod_and_leaves_prod_intact() {
    let dir = TempDir::new().unwrap();
    let (prod, stg) = setup_prod(&dir);

    // 1. sync（書込セッション開始・設計 §6.1）
    KijukuDB::replicate_db(&prod, &stg).unwrap();

    // 2. stg に LLM 編集（promote せず破棄するシナリオのシミュレート）
    {
        let stg_db = KijukuDB::open(&stg).unwrap();
        stg_db
            .create_media(&MediaInput {
                title: "stg 側の破棄される編集".to_string(),
                media_type: MediaType::Comic,
                ..Default::default()
            })
            .expect("stg edit");
    }

    // prod の状態をキャプチャ（discard 前後で「prod は一切触られない」§4.6 を検証するため）
    let prod_titles_before: Vec<String> = {
        let prod_db = KijukuDB::open(&prod).unwrap();
        (1..=3)
            .filter_map(|i| prod_db.get_media(i).map(|m| m.title))
            .collect()
    };

    // 3. discard（stg を破棄して prod から再 sync・§4.6）。処理は `replicate_db(prod, stg)` と同一。
    KijukuDB::replicate_db(&prod, &stg).unwrap();

    // 4. stg は prod で上書き復元（stg 固有の編集は破棄され prod データのみ残る）
    let stg_db = KijukuDB::open(&stg).unwrap();
    assert_eq!(stg_db.get_media(1).unwrap().title, "prod メディア1");
    assert_eq!(stg_db.get_media(2).unwrap().title, "prod メディア2");
    assert!(
        stg_db.get_media(3).is_none(),
        "discard で stg の編集は破棄される"
    );

    // 5. prod は discard 前後で無傷（§4.6・prod は一切触られない）
    let prod_titles_after: Vec<String> = {
        let prod_db = KijukuDB::open(&prod).unwrap();
        (1..=3)
            .filter_map(|i| prod_db.get_media(i).map(|m| m.title))
            .collect()
    };
    assert_eq!(
        prod_titles_before, prod_titles_after,
        "discard で prod は無傷"
    );
    assert_eq!(
        prod_titles_after,
        vec!["prod メディア1".to_string(), "prod メディア2".to_string()]
    );
}

File: ts-sdk/test/integration/sync.test.ts
/**
 * sync（prod→stg フル複製）のテスト（設計 §4.2/§4.5・TASK-50 C4）。
 *
 * `KijukuDB.replicateDb` が prod(RO)→stg(RW) の完全な複製を行うこと、
 * 既存 stg の上書きが prod で完全になること（WAL/SHM 副産物の削除込み）、
 * `src === dst` を誤設定として弾くことを検証する。
 * Rust の `rust-sdk/tests/sync_test.rs` と同等の検証を TS 側で保証する（SDK parity）。
 */
import { describe, test, expect, beforeEach, afterEach } from 'vitest';
import { KijukuDB } from '../../src/index.js';
import * as fs from 'node:fs';
import * as path from 'node:path';
import * as os from 'node:os';

describe('KijukuDB.replicateDb (sync: prod -> stg)', () => {
  let dir: string;

  beforeEach(() => {
    dir = fs.mkdtempSync(path.join(os.tmpdir(), 'kijuku-sync-'));
  });

  afterEach(() => {
    fs.rmSync(dir, { recursive: true, force: true });
  });

  test('prod(RO)->stg(RW) のフル複製がデータ・スキーマ・バージョンをコピーする', async () => {
    const prod = path.join(dir, 'kijuku.db');
    const stg = path.join(dir, 'kijuku.stg.db');

    // backup は本テストの対象外なので無効化（ノイズ回避）
    const prodDb = new KijukuDB(prod, { backup: null });
    prodDb.migrate();
    prodDb.createMedia({ title: 'prod メディア1', media_type: 'comic' });
    prodDb.createMedia({ title: 'prod メディア2', media_type: 'comic' });
    const prodVersion = prodDb.getSchemaVersion();
    prodDb.close();

    expect(fs.existsSync(stg)).toBe(false);

    await KijukuDB.replicateDb(prod, stg);

    // stg を開いて prod と一致することを検証
    expect(fs.existsSync(stg)).toBe(true);
    const stgDb = new KijukuDB(stg, { backup: null });
    expect(stgDb.getSchemaVersion()).toBe(prodVersion);
    const all = stgDb.findMedia({});
    expect(all).toHaveLength(2);
    // 順序に依存しないようソートして比較
    expect(all.map((m) => m.title).sort()).toEqual(['prod メディア1', 'prod メディア2']);
    stgDb.close();
  });

  test('既存 stg（stg 側の追加分を含む）の再 sync が prod で完全に上書きする', async () => {
    const prod = path.join(dir, 'kijuku.db');
    const stg = path.join(dir, 'kijuku.stg.db');

    const prodDb = new KijukuDB(prod, { backup: null });
    prodDb.migrate();
    prodDb.createMedia({ title: 'prod メディア1', media_type: 'comic' });
    prodDb.close();

    // 1回目の sync
    await KijukuDB.replicateDb(prod, stg);
    expect(fs.existsSync(stg)).toBe(true);

    // stg 側で独自にレコードを追加（LLM 編集のシミュレート）
    const stgDb = new KijukuDB(stg, { backup: null });
    stgDb.createMedia({
      title: 'stg 側の追加分（再 sync で消えるべき）',
      media_type: 'comic',
    });
    stgDb.close();

    // 再 sync（stg の WAL/SHM 副産物が残り得る状態からの上書き）
    await KijukuDB.replicateDb(prod, stg);

    // 再 sync 後: stg は prod と完全一致（stg 固有の追加分は削除され、prod データのみ残る）
    const stgDb2 = new KijukuDB(stg, { backup: null });
    const all = stgDb2.findMedia({});
    expect(all).toHaveLength(1);
    expect(all[0].title).toBe('prod メディア1');
    stgDb2.close();
  });

  test('src === dst は誤設定としてエラー', async () => {
    const same = path.join(dir, 'same.db');
    await expect(KijukuDB.replicateDb(same, same)).rejects.toThrow(/same path/);
  });

  test('discard（stg 破棄・再 sync・設計 §4.6）: stg を prod で上書き復元し prod は無傷', async () => {
    // discard は gate 不合格等で stg を捨てて prod から再構築する経路。処理は sync と同一
    // （`replicateDb(prod, stg)`）。ここでは discard の意味論（stg 編集の破棄 + prod 無傷）を検証。
    const prod = path.join(dir, 'kijuku.db');
    const stg = path.join(dir, 'kijuku.stg.db');

    const prodDb = new KijukuDB(prod, { backup: null });
    prodDb.migrate();
    prodDb.createMedia({ title: 'prod メディア1', media_type: 'comic' });
    prodDb.createMedia({ title: 'prod メディア2', media_type: 'comic' });
    prodDb.close();

    // 1. sync（書込セッション開始・§6.1）
    await KijukuDB.replicateDb(prod, stg);

    // 2. stg に LLM 編集（promote せず破棄するシナリオのシミュレート）
    const stgDb = new KijukuDB(stg, { backup: null });
    stgDb.createMedia({ title: 'stg 側の破棄される編集', media_type: 'comic' });
    stgDb.close();

    // prod の状態をキャプチャ（discard 前後で「prod は一切触られない」§4.6 を検証するため）
    const readProdTitles = (): string[] => {
      const db = new KijukuDB(prod, { backup: null });
      const titles = db
        .findMedia({})
        .map((m) => m.title)
        .sort();
      db.close();
      return titles;
    };
    const prodTitlesBefore = readProdTitles();

    // 3. discard（stg を破棄して prod から再 sync・§4.6）。処理は `replicateDb(prod, stg)` と同一。
    await KijukuDB.replicateDb(prod, stg);

    // 4. stg は prod で上書き復元（stg 固有の編集は破棄され prod データのみ残る）
    const stgDb2 = new KijukuDB(stg, { backup: null });
    const stgTitles = stgDb2
      .findMedia({})
      .map((m) => m.title)
      .sort();
    expect(stgTitles).toEqual(['prod メディア1', 'prod メディア2']);
    stgDb2.close();

    // 5. prod は discard 前後で無傷（§4.6・prod は一切触られない）
    const prodTitlesAfter = readProdTitles();
    expect(prodTitlesAfter).toEqual(prodTitlesBefore);
    expect(prodTitlesAfter).toEqual(['prod メディア1', 'prod メディア2']);
  });
});
